python/cudf/cudf/core/index.py

```diff
@@ -346,6 +382,18 @@ def to_arrow(self):
         """
         return self._values.to_arrow()
 
+    def tolist(self):
+        """
+        Return a list type from index data.
+
+        Returns
+        -------
+        list
+        """
+        return self.to_arrow().to_pylist()
```

Comment:
Since we decided to throw on `to_dict` we should probably do similar for `tolist`.

Reply:
Should we start doing it? I think we are using `tolist` as a means to iterate over `cudf.Series` in dask upstream. If we start poisoning `tolist` we will have to introduce `cudf` specific `to_pandas` calls in dask upstream.

For reference, we did a recent change around this: https://github.com/dask/dask/pull/6263#discussion_r433425608

or should we print a warning message in `tolist` that there is a device to host copy happening?

Reply:
Say we wanted to similarly throw on `tolist`, what are our options to handle this upstream in Dask?

Reply:
In upstream `dask` it looks like a catch 22 situation, because if we revert the changes in these [lines](https://github.com/dask/dask/blob/master/dask/dataframe/shuffle.py#L199-#L203) to create `pd.Series` instead of `cudf.Series`, we will still throw error from our `__array__` protocol as it will be an implicit conversion. 

But one alternative which I see is if we have to throw for `tolist`, we will have to add backends for each of the following functions which relies on using `tolist`
1. `set_partition`
2. `set_index`
3. `dask.utils.check_matching_columns`
4. `dask.dataframe.utils._maybe_sort`
5. `dask.dataframe.io.sql.read_sql_table`
6. `dask.dataframe.partitionquantiles.percentiles_to_weights`
7. `dask.dataframe.methods.describe_numeric_aggregate`
8. `dask.dataframe.core.repartition_npartitions`
9. `dask.dataframe.core.repartition_freq`
10. `dask.dataframe.Series.rename`

@jakirkham @rjzamora what do you think?

Reply:
cc: @shwina

Reply:
Alternatively, could we build a utility / dispatch function in dask for `tolist` and use that instead of directly calling `obj.tolist()`? Then for cuDF that utility function can do `.to_arrow().to_pylist()` for example?

Reply:
Yep, that should work actually. We'll need dispatch addition to dask upstream too so those changes will keep failing until we have dask upstream PR merged. Should this new `tolist` dispatch be added as part of this PR, or as a follow-up?

Reply:
Lets do it as a follow-up PR, but we should file issues to reference in the code before this PR is merged.

Reply:
Done